itest/lnd_invoice_migration_test.go

```diff
@@ -0,0 +1,306 @@
+		},
+	)
+
+	// Alice and bob should have one channel open with each other now.
+	ht.AssertNodeNumChannels(alice, 1)
+	ht.AssertNodeNumChannels(bob, 1)
+
+	ht.RestartNodeWithExtraArgs(bob, []string{
+		"--accept-amp",
+	})
+
+	// Step 1: Add 10 normal invoices and pay 5 of them.
+	normalInvoices := make([]*lnrpc.AddInvoiceResponse, 10)
+	for i := 0; i < 10; i++ {
+		invoice := &lnrpc.Invoice{
+			Value: int64(1000 + i*100), // Varying amounts
+			IsAmp: false,
+		}
+
+		resp := ht.Bob.RPC.AddInvoice(invoice)
+		normalInvoices[i] = resp
+	}
+
+	for _, inv := range normalInvoices {
+		sendReq := &routerrpc.SendPaymentRequest{
+			PaymentRequest: inv.PaymentRequest,
+			TimeoutSeconds: 60,
+			FeeLimitMsat:   noFeeLimitMsat,
+		}
+
+		ht.SendPaymentAssertSettled(ht.Alice, sendReq)
+	}
+
+	// Step 2: Add 10 AMP invoices and send multiple payments to 5 of them.
+	ampInvoices := make([]*lnrpc.AddInvoiceResponse, 10)
+	for i := 0; i < 10; i++ {
+		invoice := &lnrpc.Invoice{
+			Value: int64(2000 + i*200), // Varying amounts
+			IsAmp: true,
+		}
+
+		resp := bob.RPC.AddInvoice(invoice)
+		ampInvoices[i] = resp
+	}
+
+	// Select the first 5 invoices to send multiple AMP payments.
+	for i := 0; i < 5; i++ {
+		inv := ampInvoices[i]
+
+		// Send 3 payments to each.
+		for j := 0; j < 3; j++ {
+			payReq := &routerrpc.SendPaymentRequest{
+				PaymentRequest: inv.PaymentRequest,
+				TimeoutSeconds: 60,
+				FeeLimitMsat:   noFeeLimitMsat,
+				Amp:            true,
+			}
+
+			// Send a normal AMP payment first, then a spontaneous
+			// AMP payment.
+			ht.SendPaymentAssertSettled(alice, payReq)
+
+			// Generate an external payment address when attempting
+			// to pseudo-reuse an AMP invoice. When using an
+			// external payment address, we'll also expect an extra
+			// invoice to appear in the ListInvoices response, since
+			// a new invoice will be JIT inserted under a different
+			// payment address than the one in the invoice.
+			//
+			// NOTE: This will only work when the peer has
+			// spontaneous AMP payments enabled otherwise no invoice
+			// under a different payment_addr will be found.
+			payReq.PaymentAddr = ht.Random32Bytes()
+			ht.SendPaymentAssertSettled(alice, payReq)
+		}
+	}
+
+	// We can close the channel now.
+	ht.CloseChannel(alice, cp)
+
+	// Now stop Bob so we can open the DB for examination.
+	require.NoError(ht, bob.Stop())
+
+	// Open the KV channel DB.
+	db := openChannelDB(ht, bob)
+
+	query := invoices.InvoiceQuery{
+		IndexOffset: 0,
+		// As a sanity check, fetch more invoices than we have
+		// to ensure that we did not add any extra invoices.
+		NumMaxInvoices: 9999,
+	}
+
+	// Fetch all invoices and make sure we have 35 in total.
+	result1, err := db.QueryInvoices(ht.Context(), query)
+	require.NoError(ht, err)
+
+	numInvoices := len(result1.Invoices)
+
+	bob.SetExtraArgs([]string{"--db.use-native-sql"})
+
+	// Now run the migration flow three times to ensure that each run is
+	// idempotent.
```

Comment:
But I do understand it correctly that even though we run it 3 times, the migration only happens once when we start bob via the native sql flag at the beginning ?

Reply:
Yes!